src/parser.cpp

```diff
@@ -0,0 +1,637 @@
+/*
+Copyright 2015 University of Auckland
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "libcellml/parser.h"
+
+#include <libcellml/model.h>
+#include <libcellml/import.h>
+#include <libcellml/variable.h>
+#include <libcellml/component.h>
+#include <libcellml/loggererrors.h>
+
+#include "xmldoc.h"
+
+namespace libcellml {
+
+/**
+ * @brief The Parser::ParserImpl struct.
+ *
+ * The private implementation for the Parser class.
+ */
+struct Parser::ParserImpl
+{
+    Format mFormat;
+};
+
+Parser::Parser(Format format)
+    : mPimpl(new ParserImpl())
+{
+    mPimpl->mFormat = format;
+}
+
+Parser::~Parser()
+{
+    delete mPimpl;
+}
+
+Parser::Parser(const Parser& rhs)
+    : Logger(rhs)
+    , mPimpl(new ParserImpl())
+{
+    mPimpl->mFormat = rhs.mPimpl->mFormat;
+}
+
+Parser::Parser(Parser &&rhs)
+    : Logger(std::move(rhs))
+    , mPimpl(rhs.mPimpl)
+{
+    rhs.mPimpl = nullptr;
+}
+
+Parser& Parser::operator=(Parser p)
+{
+    Logger::operator =(p);
+    p.swap(*this);
+    return *this;
+}
+
+void Parser::swap(Parser &rhs)
+{
+    std::swap(this->mPimpl, rhs.mPimpl);
+}
+
+ModelPtr Parser::parseModel(const std::string &input)
+{
+    ModelPtr model = std::make_shared<Model>();
+    updateModel(model, input);
+    return model;
+}
+
+void Parser::updateModel(const ModelPtr &model, const std::string &input)
+{
+    if (mPimpl->mFormat == Format::XML) {
+        XmlDocPtr doc = std::make_shared<XmlDoc>();
+        doc->parse(input);
+        // Copy any XML parsing errors into the common error handler.
+        if (doc->xmlErrorCount() > 0) {
+            for (size_t i = 0; i < doc->xmlErrorCount(); ++i) {
+                EntityErrorPtr err = std::make_shared<EntityError>();
+                err->setDescription(doc->getXmlError(i));
+                addError(err);
+            }
+        }
+        const XmlNodePtr node = doc->getRootNode();
+        if (node->isType("model")) {
+            loadModel(model, node);
+        } else {
+            ModelErrorPtr err = std::make_shared<ModelError>();
+            err->setDescription("Unexpected XML element type (not a model).");
+            err->setModel(model);
+            addError(err);
+        }
+    }
+}
+
+void Parser::loadModel(const ModelPtr &model, const XmlNodePtr &node)
+{
+    XmlAttributePtr attribute = node->getRootAttribute();
```

Comment:
`getRootAttribute` seems a strange name: it doesn't match `getChild`, and 'root' has special meaning in some XML contexts. Would just `getAttribute` be better? Or equally change both methods to be `getFirstAttribute` and `getFirstChild` for clarity?